### DISCUSSION TITLE
Auto leave rooms and authentication.

### DISCUSSION THREAD

**Discussion by jonataslaw** [uNetworking/uWebSockets#985] [Q&A] [answered]
Hi, thanks for this library.
I have an application with about 2 million users, and I am looking for alternatives to socket.io to reduce latency, and if possible save resources. I have seen some articles about this library and am amazed by the benchmarks and tests.
However, with socket-io there are two things I didn't find in doc if it is possible to do with this library.
1- When a user is disconnected, with socket.io automatically is removed from the rooms, and no message is sent to the disconnected socket. Is this the same with this library? And what happens when the connection goes down? Are there attempts to reconnect, or should this be done by the client?
2- I use authentication with database and socket.io, the user is authenticated, receives a token, and if his token is valid, his socket receives a user_id. Is it possible to use authentication with this library? If somehow an unidentified user could post to rooms / subscribes unauthorized I would be ruined.
Thanks for your time, your work is amazing!

**Comment by ghost** [comment, marked as answer]
1, yes but you can also just call unsubscribeAll but yes it does auto remove disconnected sockets from rooms (here called topics)

2, yes the WebSocket has its http request passed in upgrade so you can do auth that way.

Reconnect you have to solve yourself but that's like 9 lines of JavaScript

**Comment by ghost** [comment]
#986 I apparently forgot to add that function.
